File: Top_Interview_Questions/validParenthesis.go
func isValid(s string) bool {
    var stack []rune;
    opening := map[rune]rune{')': '(', '}': '{', ']': '['};
    for _, char := range s {
        opening_char, found := opening[char];
        if found {
            if len(stack) == 0 || stack[len(stack) - 1] != opening_char {
                return false;
            } else {
                stack = stack[:len(stack) - 1]; // pop
            }
        } else {
            stack = append(stack, char); // push
        }
    }
    return len(stack) == 0;
}

File: NeetCode_Roadmap/valid_parenthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        open2close = {
            "(": ")", "[": "]", "{": "}"
        }
        stack = []
        for c in s:
            if c in open2close.keys():
                stack.append(c)
            else:
                if len(stack) == 0 or open2close[stack[-1]] != c:
                    return False
                stack.pop()

        return len(stack) == 0
                    
